core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -651,60 +651,126 @@ class UserService(protected val ctx: RawlsRequestContext,
+    }
+
+  private def addUserToBillingProjectInner(projectName: RawlsBillingProjectName,
+                                           projectAccessUpdate: ProjectAccessUpdate,
+                                           policies: Seq[SamResourcePolicyName]
+  ): Future[Unit] =
+    for {
+      _ <- Future.traverse(policies) { policy =>
+        samDAO
+          .addUserToPolicy(SamResourceTypeNames.billingProject,
+                           projectName.value,
+                           policy,
+                           projectAccessUpdate.email,
+                           ctx.userInfo
+          )
+          .recoverWith { case regrets: Throwable =>
+            if (policy == SamBillingProjectPolicyNames.canComputeUser) {
+              logger.info(
+                s"error adding user to canComputeUser policy for $projectName likely because it is a v2 billing project which does not have a canComputeUser policy. regrets: ${regrets.getMessage}"
+              )
+              Future.successful(())
+            } else {
+              Future.failed(regrets)
+            }
+          }
       }
+    } yield {}
 
+  def addUserToBillingProjectV2(projectName: RawlsBillingProjectName,
+                                projectAccessUpdate: ProjectAccessUpdate
+  ): Future[Unit] =
+    requireProjectAction(projectName, SamBillingProjectActions.alterPolicies) {
       for {
-        _ <- Future.traverse(policies) { policy =>
-          samDAO
-            .addUserToPolicy(SamResourceTypeNames.billingProject,
-                             projectName.value,
-                             policy,
-                             projectAccessUpdate.email,
-                             ctx.userInfo
+        billingProjectId <- getBillingProfileId(projectName)
+        policies = billingProjectId match {
+          case None => getLegacyBillingPolicies(projectAccessUpdate.role)
+          case Some(billingProjectId) =>
+            billingProfileManagerDAO.addProfilePolicyMember(
+              UUID.fromString(billingProjectId),
+              projectAccessUpdate.role,
+              projectAccessUpdate.email,
+              ctx
             )
-            .recoverWith { case regrets: Throwable =>
-              if (policy == SamBillingProjectPolicyNames.canComputeUser) {
-                logger.info(
-                  s"error adding user to canComputeUser policy for $projectName likely because it is a v2 billing project which does not have a canComputeUser policy. regrets: ${regrets.getMessage}"
-                )
-                Future.successful(())
-              } else {
-                Future.failed(regrets)
-              }
-            }
+            Seq(getV2BillingPolicy(projectAccessUpdate.role))
         }
+        _ <- addUserToBillingProjectInner(projectName, projectAccessUpdate, policies)
       } yield {}
     }
 
+  def getBillingProfileId(projectName: RawlsBillingProjectName): Future[Option[String]] =
```

Comment:
This would be a good method to put on the [billing project repository.](https://github.com/broadinstitute/rawls/blob/develop/core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BillingRepository.scala)

Reply:
I gave this a shot in https://github.com/broadinstitute/rawls/pull/2046/commits/7d22a12f5fb2e29e6024a4b25a7a784db2d0e6ba-- let me know what you think.

Reply:
This is causing a lot of tests to bomb because I need to provide an implementation for `getBillingProfileId`. If you think this is worth pursuing, I will probably need some help tomorrow figuring out how best to do this.

Reply:
Actually, I realize now that my approach of passing BillingRegistry into UserService is overkill given the current situation where UserService directly uses the `slickDataSource` for all other database access. I think I'm going to revert these changes and instead simply create an instance of BillingRegistry in UserService with the `slickDataSource` reference it already has. That will drastically reduce the changes, and once/if UserService has migrated to using BillingRegistry for everything we can start passing it into UserService in place of the `slickDataSource`. Sound good to you, @aherbst-broad?

Reply:
@cahrens ah, yeah, that makes sense. Thanks for the explanation. (assuming you mean BillingRepository instead of BillingRegistry).